base/base/defines.h

```diff
@@ -108,16 +108,22 @@
         {
             [[noreturn]] void abortOnFailedAssertion(const String & description);
         }
-        #define chassert(x) do { static_cast<bool>(x) ? void(0) : ::DB::abortOnFailedAssertion(#x); } while (0)
+        #define chassert_1(x, ...) do { static_cast<bool>(x) ? void(0) : ::DB::abortOnFailedAssertion(#x); } while (0)
+        #define chassert_2(x, comment, ...) do { static_cast<bool>(x) ? void(0) : ::DB::abortOnFailedAssertion(comment); } while (0)
         #define UNREACHABLE() abort()
         // clang-format off
     #else
         /// Here sizeof() trick is used to suppress unused warning for result,
         /// since simple "(void)x" will evaluate the expression, while
         /// "sizeof(!(x))" will not.
-        #define chassert(x) (void)sizeof(!(x))
+        #define chassert_1(x, ...) (void)sizeof(!(x))
+        #define chassert_2(x, comment, ...) (void)sizeof(!(x))
         #define UNREACHABLE() __builtin_unreachable()
     #endif
+        #define CHASSERT_DISPATCH(_1,_2, N,...) N(_1, _2)
+        #define CHASSERT_INVOKE(tuple) CHASSERT_DISPATCH tuple
+        #define chassert(...) CHASSERT_INVOKE((__VA_ARGS__, chassert_2, chassert_1))
```

Comment:
For the future, in case I come back to try to understand this:
* When called with one argument, CHASSERT_INVOKE is called with a tuple (argument, chassert_2, chassert_1), so the dispatch calls to chassert_1(argument, chassert_2) and chassert_2 is ignored.
* When called with 2 arguments, CHASSERT_INVOKE is called with a tuple (arg1, arg2, chassert_2, chassert_1) so the dispatch calls chassert_2(arg1, arg2).